// @loopback/http-server/dist8/src/http-server.js
"use strict";
// Copyright IBM Corp. 2017,2018. All Rights Reserved.
// Node module: @loopback/http-server
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT
Object.defineProperty(exports, "__esModule", { value: true });
const http_1 = require("http");
const pEvent = require("p-event");
/**
 * HTTP / HTTPS server used by LoopBack's RestServer
 *
 * @export
 * @class HttpServer
 */
class HttpServer {
    /**
     * @param httpServerOptions
     * @param httpRequestListener
     */
    constructor(httpRequestListener, httpServerOptions) {
        this._listening = false;
        this.httpRequestListener = httpRequestListener;
        if (!httpServerOptions)
            httpServerOptions = {};
        this._port = httpServerOptions.port || 0;
        this._host = httpServerOptions.host || undefined;
        this._protocol = httpServerOptions.protocol || 'http';
    }
    /**
     * Starts the HTTP / HTTPS server
     */
    async start() {
        this.httpServer = http_1.createServer(this.httpRequestListener);
        this.httpServer.listen(this._port, this._host);
        await pEvent(this.httpServer, 'listening');
        this._listening = true;
        this._address = this.httpServer.address();
    }
    /**
     * Stops the HTTP / HTTPS server
     */
    async stop() {
        if (this.httpServer) {
            this.httpServer.close();
            await pEvent(this.httpServer, 'close');
            this._listening = false;
        }
    }
    /**
     * Protocol of the HTTP / HTTPS server
     */
    get protocol() {
        return this._protocol;
    }
    /**
     * Port number of the HTTP / HTTPS server
     */
    get port() {
        return (this._address && this._address.port) || this._port;
    }
    /**
     * Host of the HTTP / HTTPS server
     */
    get host() {
        return (this._address && this._address.address) || this._host;
    }
    /**
     * URL of the HTTP / HTTPS server
     */
    get url() {
        let host = this.host;
        if (this._address.family === 'IPv6') {
            host = `[${host}]`;
        }
        return `${this._protocol}://${host}:${this.port}`;
    }
    /**
     * State of the HTTP / HTTPS server
     */
    get listening() {
        return this._listening;
    }
    /**
     * Address of the HTTP / HTTPS server
     */
    get address() {
        return this._listening ? this._address : undefined;
    }
}
exports.HttpServer = HttpServer;
